### DISCUSSION TITLE
Why do "use client" components run off the client?

### DISCUSSION THREAD

**Discussion by tinleym** [vercel/next.js#43270] [Help] [answered]
In dev mode, I'm seeing logs from a "use client" component in my server stdout... I thought there was a separation between the two?

I think these two client environments are the cause of some unexpected behaviors. 

I have a client data store that doesn't always get cleared on refresh in this non-browser client environment (as seen through my server stdout), but does in the browser. When I delete a cookie in dev tools, I get mismatched data and a hydration error. 

This non-browser client environment also persists across different browser instances; the same store behavior and hydration error crops up when I switch back and forth between Chrome and Safari.

Said store does get cleared if I make code change, or if I kill and restart the dev process, and the error goes away.

I'm guessing this has something to do with HMR, and won't apply to production. However, it would be nice to know :)

**Comment by icyJoseph** [comment, marked as answer]
The term Client Component, as opposed to Server Component, is unfortunately slightly misleading. 

> It is not only in the client, but rather, also in the client

The so called Client Components, are the components we have been writing up to this point. During SSR, React walks the tree, collecting HTML, roughly speaking, when it finds either client or server components, through different mechanisms, it collects the JSX for these. That collection of React Elements is stringified into HTML, and sent to the client. Simplifying a lot here, but that's the sketch of the process.

What's so special about Server Components is that their JS doesn't need to be shipped to the browser. Client Components, or just the regular components we've been writing up to this point, do use JS to drive interactivity, so their JS is sent to the client. Again simplifying here for sake of clarity.

> If you were to use unguarded `window` access within a client component, you'd also see error logs on the server.

Another way to think of this, imagine you are an SEO bot, or crawler, if you arrive to my page, I want you to see all of the HTML, even that of client components, because the fact that such content is interactive doesn't equal that it is irrelevant, it must be there, so I want the SSR process to touch the client components.

So to sum up, Server Components are server only, Client Components are also shared to the client.

Regarding your problem I am not sure, but it sounds like hydration mismatch because the HTML the server is sending doesn't match up with the hydration process result. Possibly because your client components crash or error out during SSR.

**Reply by tinleym** [reply]
@icyJoseph this is really helpful, thank you!

The thing that's still strange about my particular problem is that I have a client store that persists outside the browser, surviving and mutating between refreshes and across browser instances. It's not just that the data gets stale; data from the current instance is shared with data from past sessions and other instances.

I don't see where this kind of client store persistence could live in production, so I'm pretty sure it's an HMR quirk... are there any other possibilities you can think of?

**Reply by mrbjjackson** [reply]
@icyJoseph Your answer here has answered some questions I've got about server components vs client components. Thanks very much.

Please can you just confirm something for me to check I've understood correctly:

I use `framer-motion` a lot to animate elements on my pages. As far as I can see, `framer-motion` is incompatible with server components (the errors I've seen suggest it's because `motion` components use React Context which isn't compatible with server components). So this requires me to mark any pages that include `framer-motion` as `"use client"`.

I was worried that doing so would mean search engines would no longer see the contents of these components. But from what you've written above, it seems that, these components _would_ still be SSR'd - it's just they would _also_ ship a load of javascript for the `motion` functionality. Have I got that right?

Thanks in advance,
Ben

**Reply by icyJoseph** [reply]
> I don't see where this kind of client store persistence could live in production, so I'm pretty sure it's an HMR quirk... are there any other possibilities you can think of?

@tinleym  Do you think there's to some capacity, a service worker involved? Service workers with stale while revalidate, https://developer.chrome.com/docs/workbox/caching-strategies-overview/#stale-while-revalidate, which is not very good for dynamic data of course, an infamous for forcing people to refresh the page to see the latest data.

Note that, AFAIK, a bad deploy on a service worker will keep it alive, even if you have moved on to a whole new architecture... https://twitter.com/sanketsahu/status/1133223194888773632

> I use framer-motion a lot to animate elements on my pages. As far as I can see, framer-motion is incompatible with server components (the errors I've seen suggest it's because motion components use React Context which isn't compatible with server components). So this requires me to mark any pages that include framer-motion as "use client".
> 
> I was worried that doing so would mean search engines would no longer see the contents of these components. But from what you've written above, it seems that, these components would still be SSR'd - it's just they would also ship a load of javascript for the motion functionality. Have I got that right?
> 

@mrbjjackson  More or less yes. You can test this yourself, disable JS, or use Postman or [VSCode Rest Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client), to make a request as a "bot/crawler/seo", entity and see what you receive 😄 

Here's a response sequence  you might want to read through https://github.com/vercel/next.js/discussions/41745#discussioncomment-4042599

Though I've tried to get a statement from someone in the Vercel team, I've not been able to, that is, regarding the actual scope of client components within Next 13's architecture, not only this one: https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md

**Reply by tinleym** [reply]
@icyJoseph that's a good theory! I'm using Relay, which doesn't mention using Service Workers, but at least has a similar cache-network fetching process that a Service Worker might have. The thing is, I'm seeing this persistent shared state between Chrome and Safari -- if I fetch from Chrome using one cookie, and then fetch from Safari using another, I'll initially get the Chrome result server-rendered in Safari. This leads me to believe they're both reading from common local files.

**Reply by icyJoseph** [reply]
Do you see this in development mode as well?

**Reply by tinleym** [reply]
I've only seen it in dev mode. Which makes me think I probably shouldn't be concerned about it, but I'd feel better if I knew what was going on, if it's a known and accepted behavior, etc. etc.
